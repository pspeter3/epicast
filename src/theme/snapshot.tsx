import * as React from "react";
import TestRenderer from "react-test-renderer";

export const snapshot = (name: string, SFC: React.SFC<{}>): void => {
    describe(name, () => {
        it(`should match snapshot`, () => {
            const component = TestRenderer.create(<SFC>Snapshot</SFC>);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
};
