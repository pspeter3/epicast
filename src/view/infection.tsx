import * as React from "react";
import { BottomButton, IconButton } from "../theme/buttons";
import { CheckCircleIcon, CircleIcon, TargetIcon } from "../theme/icons";
import { Main, Row, Section } from "../theme/layout";
import { Text } from "../theme/typography";

export const Infection: React.SFC<{}> = () => {
    return (
        <React.Fragment>
            <Main>
                <Section>
                    <Row>
                        <IconButton>
                            <CheckCircleIcon />
                        </IconButton>
                        <Text>Atlanta</Text>
                    </Row>
                </Section>
                <Section>
                    <Row>
                        <IconButton>
                            <CircleIcon />
                        </IconButton>
                        <Text>Washington D.C.</Text>
                        <IconButton>
                            <TargetIcon />
                        </IconButton>
                    </Row>
                    <Row>
                        <IconButton>
                            <CircleIcon />
                        </IconButton>
                        <Text>Washington D.C</Text>
                        <IconButton>
                            <TargetIcon />
                        </IconButton>
                    </Row>
                </Section>
            </Main>
            <BottomButton>Skip Infection</BottomButton>
        </React.Fragment>
    );
};
