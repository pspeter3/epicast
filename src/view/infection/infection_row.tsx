import * as React from "react";
import { IconButton } from "../../theme/buttons";
import { TargetIcon } from "../../theme/icons";
import { Checkbox } from "../../theme/inputs";
import { Row } from "../../theme/layout";
import { Text } from "../../theme/typography";

export interface Props {
    city: string;
    checked?: boolean;
    onToggle: (city: string) => void;
    onEpidemic?: (city: string) => void;
}

export class InfectionRow extends React.PureComponent<Props, {}> {
    public render() {
        const { city, checked, onEpidemic } = this.props;
        return (
            <Row>
                <Checkbox aria-checked={checked} onClick={this._onToggle} />
                <Text>{city}</Text>
                {onEpidemic && (
                    <IconButton onClick={this._onEpidemic}>
                        <TargetIcon />
                    </IconButton>
                )}
            </Row>
        );
    }

    private _onToggle = () => {
        this.props.onToggle(this.props.city);
    };

    private _onEpidemic = () => {
        this.props.onEpidemic!(this.props.city);
    };
}
