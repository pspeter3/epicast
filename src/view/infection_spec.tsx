import * as React from "react";
import { snapshotSuite } from "../util/snapshot";
import { Infection } from "./infection";

snapshotSuite("Infection", () => <Infection />);
