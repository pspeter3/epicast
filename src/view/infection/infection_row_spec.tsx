import { shallow } from "enzyme";
import * as React from "react";
import { IconButton } from "../../theme/buttons";
import { Checkbox } from "../../theme/inputs";
import { snapshot } from "../../util/snapshot";
import { InfectionRow } from "./infection_row";

describe("InfectionRow", () => {
    describe("snapshots", () => {
        it(
            "should match default snapshot",
            snapshot(() => <InfectionRow city="Test" onToggle={jest.fn()} />),
        );
        it(
            "should match checked snapshot",
            snapshot(() => <InfectionRow city="Test" onToggle={jest.fn()} />),
        );
        it(
            "should match epidemic snapshot",
            snapshot(() => (
                <InfectionRow
                    city="Test"
                    checked={false}
                    onToggle={jest.fn()}
                    onEpidemic={jest.fn()}
                />
            )),
        );
    });

    describe("handlers", () => {
        it("should trigger change", () => {
            const city = "San Francisco";
            const onToggle = jest.fn();
            const tree = shallow(<InfectionRow city={city} onToggle={onToggle} />);
            tree.find(Checkbox).simulate("click");
            expect(onToggle).toHaveBeenCalledWith(city);
        });

        it("should trigger epidemic", () => {
            const city = "San Francisco";
            const onEpidemic = jest.fn();
            const tree = shallow(
                <InfectionRow city={city} onToggle={jest.fn()} onEpidemic={onEpidemic} />,
            );
            tree.find(IconButton).simulate("click");
            expect(onEpidemic).toHaveBeenCalledWith(city);
        });
    });
});
