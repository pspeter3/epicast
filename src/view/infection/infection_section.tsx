import * as React from "react";
import { Stack } from "../../core/types";
import { Section } from "../../theme/layout";
import { value } from "../../util/stacks";
import { InfectionRow } from "./infection_row";

export interface Props {
    index: number;
    stack: Stack;
    checked: Stack;
    onToggle: (index: number, city: string) => void;
    onEpidemic?: (city: string) => void;
}

export class InfectionSection extends React.PureComponent<Props, {}> {
    public render() {
        const { stack, checked, onEpidemic } = this.props;
        return (
            <Section>
                {Object.keys(stack)
                    .sort()
                    .reduce(
                        (children, city) => {
                            const count = stack[city];
                            const numChecked = value(checked, city);
                            for (let i = 0; i < count; i++) {
                                children.push(
                                    <InfectionRow
                                        key={`${city}${i}`}
                                        city={city}
                                        checked={i < numChecked}
                                        onToggle={this._onToggle}
                                        onEpidemic={onEpidemic}
                                    />,
                                );
                            }
                            return children;
                        },
                        [] as Array<React.ReactElement<{}>>,
                    )}
            </Section>
        );
    }

    private _onToggle = (city: string) => {
        this.props.onToggle(this.props.index, city);
    };
}
