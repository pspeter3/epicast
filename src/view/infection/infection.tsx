import * as React from "react";
import { Deck } from "../../core/types";
import { PrimaryButton } from "../../theme/buttons";
import { Main } from "../../theme/layout";
import { size, union, unit } from "../../util/stacks";
import { InfectionSection } from "./infection_section";

export interface Props {
    deck: Deck;
    onInfect: (deck: Deck) => void;
    onEpidemic: (city: string) => void;
}

export interface State {
    selected: Deck;
}

export class Infection extends React.PureComponent<Props, State> {
    public state: State = { selected: [] };

    public render() {
        const { deck, onEpidemic } = this.props;
        const { selected } = this.state;
        const count = selected.reduce((sum, stack) => sum + size(stack), 0);
        return (
            <Main>
                {deck.reduceRight(
                    (children, stack, index) => {
                        children.push(
                            <InfectionSection
                                key={index}
                                index={index}
                                stack={stack}
                                checked={index >= selected.length ? {} : selected[index]}
                                onToggle={this._onToggle}
                                onEpidemic={index === 0 ? onEpidemic : undefined}
                            />,
                        );
                        return children;
                    },
                    [] as Array<React.ReactElement<{}>>,
                )}
                <PrimaryButton onClick={this._onInfect}>
                    {count === 0 ? "Skip Infection" : "Infect"}
                </PrimaryButton>
            </Main>
        );
    }

    private _onInfect = () => {
        this.props.onInfect(this.state.selected);
        this.setState({ selected: [] });
    };

    private _onToggle = (index: number, city: string) => {
        const delta = unit(city);
        let selected: Deck = this.state.selected.map((stack, i) => {
            return i === index ? union(stack, delta) : stack;
        });
        for (let i = selected.length; i <= index; i++) {
            selected = i === index ? selected.concat(delta) : selected.concat({});
        }
        this.setState({ selected });
    };
}
