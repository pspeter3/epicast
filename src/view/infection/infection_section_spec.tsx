import { shallow } from "enzyme";
import * as React from "react";
import { snapshot } from "../../util/snapshot";
import { InfectionRow } from "./infection_row";
import { InfectionSection } from "./infection_section";

const PRIMARY = "San Francisco";
const SECONDARY = "Los Angeles";

describe("InfectionSection", () => {
    describe("snapshots", () => {
        it(
            "should match snapshot with checked cities",
            snapshot(() => (
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{ [PRIMARY]: 1 }}
                    onToggle={jest.fn()}
                />
            )),
        );
        it(
            "should match snapshot with epidemic",
            snapshot(() => (
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{}}
                    onToggle={jest.fn()}
                    onEpidemic={jest.fn()}
                />
            )),
        );
    });

    describe("handlers", () => {
        it("should handle toggle", () => {
            const onToggle = jest.fn();
            const tree = shallow(
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{}}
                    onToggle={onToggle}
                />,
            );
            tree
                .find(InfectionRow)
                .first()
                .props().onToggle!(PRIMARY);
            expect(onToggle).toHaveBeenCalledWith(0, PRIMARY);
        });
    });
});
