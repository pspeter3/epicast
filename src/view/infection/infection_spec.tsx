import { shallow } from "enzyme";
import * as React from "react";
import { BottomButton } from "../../theme/buttons";
import { snapshot } from "../../util/snapshot";
import { Infection } from "./infection";
import { InfectionSection } from "./infection_section";

const PRIMARY = "San Francisco";
const SECONDARY = "Los Angeles";

describe("Infection", () => {
    describe("snapshots", () => {
        it(
            "should match snapshot",
            snapshot(() => (
                <Infection
                    deck={[{ [PRIMARY]: 1 }, { [PRIMARY]: 2, [SECONDARY]: 2 }]}
                    onInfect={jest.fn()}
                    onEpidemic={jest.fn()}
                />
            )),
        );

        it("should match selected snapshot", () => {
            const tree = shallow(
                <Infection
                    deck={[{ [PRIMARY]: 1 }, { [PRIMARY]: 2, [SECONDARY]: 2 }]}
                    onInfect={jest.fn()}
                    onEpidemic={jest.fn()}
                />,
            );
            tree
                .find(InfectionSection)
                .first()
                .prop("onToggle")(1, PRIMARY);
            expect(tree).toMatchSnapshot();
        });
    });

    describe("handlers", () => {
        it("should handle skipping infection", () => {
            const onInfect = jest.fn();
            const tree = shallow(
                <Infection
                    deck={[{ [PRIMARY]: 1 }, { [PRIMARY]: 2, [SECONDARY]: 2 }]}
                    onInfect={onInfect}
                    onEpidemic={jest.fn()}
                />,
            );
            tree.find(BottomButton).simulate("click");
            expect(onInfect).toHaveBeenCalledWith([]);
        });

        it("should handle selection", () => {
            const onInfect = jest.fn();
            const tree = shallow(
                <Infection
                    deck={[{ [PRIMARY]: 1 }, { [PRIMARY]: 2, [SECONDARY]: 2 }]}
                    onInfect={onInfect}
                    onEpidemic={jest.fn()}
                />,
            );
            tree
                .find(InfectionSection)
                .first()
                .prop("onToggle")(1, PRIMARY);
            tree
                .find(InfectionSection)
                .last()
                .prop("onToggle")(0, PRIMARY);
            tree.find(BottomButton).simulate("click");
            expect(onInfect).toHaveBeenCalledWith([{ [PRIMARY]: 1 }, { [PRIMARY]: 1 }]);
            expect(tree.state("selected")).toEqual([]);
        });
    });
});
