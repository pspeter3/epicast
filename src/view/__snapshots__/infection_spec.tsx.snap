// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infection should match snapshot 1`] = `
Array [
  <main
    className="main"
  >
    <section
      className="section"
    >
      <div
        className="row"
      >
        <button
          className="icon-button"
        >
          <svg
            className="feather feather-check-circle"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
        </button>
        <p
          className="text"
        >
          Atlanta
        </p>
      </div>
    </section>
    <section
      className="section"
    >
      <div
        className="row"
      >
        <button
          className="icon-button"
        >
          <svg
            className="feather feather-circle"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </button>
        <p
          className="text"
        >
          Washington D.C.
        </p>
        <button
          className="icon-button"
        >
          <svg
            className="feather feather-target"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <circle
              cx="12"
              cy="12"
              r="6"
            />
            <circle
              cx="12"
              cy="12"
              r="2"
            />
          </svg>
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="icon-button"
        >
          <svg
            className="feather feather-circle"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </button>
        <p
          className="text"
        >
          Washington D.C
        </p>
        <button
          className="icon-button"
        >
          <svg
            className="feather feather-target"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <circle
              cx="12"
              cy="12"
              r="6"
            />
            <circle
              cx="12"
              cy="12"
              r="2"
            />
          </svg>
        </button>
      </div>
    </section>
  </main>,
  <button
    className="bottom-button"
  >
    Skip Infection
  </button>,
]
`;
